From 2db4395bf193b40cde46d87bd7e7fc58d3554763 Mon Sep 17 00:00:00 2001
From: Samuel Hym <samuel@tarides.com>
Date: Tue, 3 Sep 2024 15:18:23 +0200
Subject: [PATCH 09/21] Enable bootstrapping flexdll in the cross-compiler
 setting

Add a rule to build flexdll in the cross-compiler setting, namely
building flexdll on Unix, by driving its `Makefile` so that Windows
resources are not built and only the .opt version is really built (and
copied to the byte binary directory nevertheless)
Use the `-o` flag to tell `make` to never try to rebuild `flexlink` (as
it would otherwise, according to the dependencies in the main
`Makefile`)
---
 Makefile.cross | 55 ++++++++++++++++++++++++++++++++++++++------------
 1 file changed, 42 insertions(+), 13 deletions(-)

diff --git a/Makefile.cross b/Makefile.cross
index c30d2e2b4f..e01792de20 100644
--- a/Makefile.cross
+++ b/Makefile.cross
@@ -32,32 +32,61 @@ CROSS_COMPILER_OVERRIDES=$(CROSS_OVERRIDES) CAMLC=ocamlc CAMLOPT=ocamlopt \
 CROSS_COMPILERLIBS_OVERRIDES=$(CROSS_OVERRIDES) CAMLC=ocamlc \
   CAMLOPT="$(ROOTDIR)/ocamlopt.opt$(EXE) $(STDLIBFLAGS)"
 
+ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
+# Declare flexlink to be an 'old' file, so that make doesn't try to rebuild it
+# with the build rules in `Makefile`; its build is driven by the `cross-flexdll`
+# recipe provided here instead
+OLDS := -o $(BYTE_BINDIR)/flexlink$(EXE)
+else
+OLDS :=
+endif
+
 # The compiler libs that should be rebuilt for target (they are first built for
 # host as part of the .opt compilers)
 CROSSCOMPILERLIBS := $(addprefix compilerlibs/,$(addsuffix .cmxa,\
     ocamlcommon ocamlmiddleend ocamlbytecomp ocamloptcomp ocamltoplevel))
 
 .PHONY: crossopt
+ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
+crossopt: cross-flexdll
+	$(MAKE) runtime-all $(OLDS)
+else
+# In that case, $(OLDS) is empty, we can depend directly on runtime-all
 crossopt: runtime-all
+endif
 	$(MAKE) ocamlc $(TOOLS_BYTECODE_TARGETS) expunge$(EXE) \
-	  $(CROSS_COMPILER_OVERRIDES)
-	$(MAKE) library $(CROSS_OVERRIDES)
-	$(MAKE) ocamlyacc $(CROSS_OVERRIDES)
-	$(MAKE) ocamllex $(CROSS_COMPILER_OVERRIDES)
-	$(MAKE) ocaml $(CROSS_COMPILER_OVERRIDES)
-	$(MAKE) dynlink-all $(CROSS_OVERRIDES)
-	$(MAKE) -C otherlibs all $(CROSS_OVERRIDES)
-	$(MAKE) runtimeopt
+	  $(CROSS_COMPILER_OVERRIDES) $(OLDS)
+	$(MAKE) library $(CROSS_OVERRIDES) $(OLDS)
+	$(MAKE) ocamlyacc $(CROSS_OVERRIDES) $(OLDS)
+	$(MAKE) ocamllex $(CROSS_COMPILER_OVERRIDES) $(OLDS)
+	$(MAKE) ocaml $(CROSS_COMPILER_OVERRIDES) $(OLDS)
+	$(MAKE) dynlink-all $(CROSS_OVERRIDES) $(OLDS)
+	$(MAKE) -C otherlibs all $(CROSS_OVERRIDES) $(OLDS)
+	$(MAKE) runtimeopt $(OLDS)
 	$(MAKE) ocamlc.opt ocamlopt.opt $(TOOLS_NATIVE_TARGETS) \
-	  $(CROSS_COMPILER_OVERRIDES) "$(HOST_ZSTD_LIBS)"
-	$(MAKE) libraryopt $(CROSS_OVERRIDES)
-	$(MAKE) otherlibrariesopt ocamltoolsopt $(CROSS_OVERRIDES)
-	$(MAKE) tools-allopt.opt $(CROSS_COMPILER_OVERRIDES)
+	  $(CROSS_COMPILER_OVERRIDES) "$(HOST_ZSTD_LIBS)" $(OLDS)
+	$(MAKE) libraryopt $(CROSS_OVERRIDES) $(OLDS)
+	$(MAKE) otherlibrariesopt ocamltoolsopt $(CROSS_OVERRIDES) $(OLDS)
+	$(MAKE) tools-allopt.opt $(CROSS_COMPILER_OVERRIDES) $(OLDS)
 	# We now build the compiler libs again, but for target this time
 	rm -f $(ocamlcommon_NCOBJS) $(ocamlmiddleend_NCOBJS) \
 	  $(ocamlbytecomp_NCOBJS) $(ocamloptcomp_NCOBJS) \
 	  $(ocamltoplevel_NCOBJS) $(CROSSCOMPILERLIBS)
-	$(MAKE) $(CROSSCOMPILERLIBS) $(CROSS_COMPILERLIBS_OVERRIDES)
+	$(MAKE) $(CROSSCOMPILERLIBS) $(CROSS_COMPILERLIBS_OVERRIDES) $(OLDS)
+
+.PHONY: cross-flexdll
+cross-flexdll: | $(BYTE_BINDIR) $(OPT_BINDIR)
+	rm -f $(FLEXDLL_SOURCE_DIR)/flexlink.exe
+	$(MAKE) -C $(FLEXDLL_SOURCE_DIR) $(FLEXLINK_BUILD_ENV) \
+	  NATDYNLINK=false LINKFLAGS= flexlink.exe support
+	$(LN) $(FLEXDLL_SOURCE_DIR)/flexlink.exe flexlink.opt.exe
+	$(LN) flexlink.opt.exe flexlink.byte.exe
+	cp flexlink.byte.exe $(BYTE_BINDIR)/flexlink
+	cd $(BYTE_BINDIR) && $(LN) flexlink flexlink.exe
+	cp $(addprefix $(FLEXDLL_SOURCE_DIR)/, $(FLEXDLL_OBJECTS)) $(BYTE_BINDIR)
+	cp flexlink.opt.exe $(OPT_BINDIR)/flexlink
+	cd $(OPT_BINDIR) && $(LN) flexlink flexlink.exe
+	cp $(addprefix $(FLEXDLL_SOURCE_DIR)/, $(FLEXDLL_OBJECTS)) $(OPT_BINDIR)
 
 INSTALL_OVERRIDES=build_ocamldoc=false WITH_DEBUGGER= OCAMLRUN=ocamlrun
 
-- 
2.47.2

